pretyping/recordops.ml

```diff
@@ -367,3 +367,52 @@ let is_open_canonical_projection env sigma (c,args) =
       not (isConstruct sigma hd)
     with Failure _ -> false
   with Not_found -> false
+
+(** Global data for record encodings *)
+
+let record_table =
+  Summary.ref ~name:"RECORD-EMULATION-TABLE" Mindset_env.empty
+
+(** In order to properly handle functors, we need to implement a libobject which
+    is at the same time `Keep` and substitutive. We emulate that by using a set
+    of MutInd.t that we duplicate at each substitution. The first instance
+    stands for the kept objects, and we apply the substitution to the second
+    instance to treat them as substitutive. *)
+
+let cache_record (_, minds) =
+  let fold mind accu = Mindset_env.add mind accu in
+  record_table := Mindset_env.fold fold minds !record_table
```

Comment:
Isn't there a union function for this?

Reply:
More efficient to do this when we know that one side is generally smaller than the other.